mlflow/models/signature.py

```diff
@@ -301,38 +301,50 @@ def _extract_type_hints(f, input_arg_index):
     return _TypeHints(hints.get(arg_name), hints.get("return"))
 
 
-def _infer_signature_from_type_hints(func, input_arg_index, input_example=None):
+def _infer_signature_from_type_hints(
+    func, input_arg_index, input_example=None, params_example=None
+):
     hints = _extract_type_hints(func, input_arg_index)
     if hints.input is None:
         return None
 
     input_schema = _infer_schema_from_type_hint(hints.input, input_example) if hints.input else None
-    output_example = func(input_example) if input_example else None
+    params_schema = _infer_param_schema(params_example) if params_example else None
+    if params_example and "params" in inspect.signature(func).parameters:
+        output_example = func(input_example, params=params_example) if input_example else None
+    else:
+        output_example = func(input_example) if input_example else None
     output_schema = (
         _infer_schema_from_type_hint(hints.output, output_example) if hints.output else None
     )
-    if input_schema is None and output_schema is None:
+    if not any([input_schema, output_schema, params_schema]):
         return None
-    return ModelSignature(inputs=input_schema, outputs=output_schema)
+    return ModelSignature(inputs=input_schema, outputs=output_schema, params=params_schema)
 
 
 def _infer_signature_from_input_example(
-    input_example: ModelInputExample, wrapped_model
+    input_example: ModelInputExample, wrapped_model, params_example: Optional[Dict[str, Any]] = None
 ) -> Optional[ModelSignature]:
     """
     Infer the signature from an example input and a PyFunc wrapped model. Catches all exceptions.
 
     :param input_example: An instance representing a typical input to the model.
     :param wrapped_model: A PyFunc wrapped model which has a `predict` method.
+    :param params_example: A dictionary of example parameters for model inference.
     :return: A `ModelSignature` object containing the inferred schema of both the model's inputs
         based on the `input_example` and the model's outputs based on the prediction from the
         `wrapped_model`.
     """
     try:
-        input_ex = _Example(input_example).inference_data
-        input_schema = _infer_schema(input_ex)
+        # We shouldn't do this at least for lists --> it produces wrong results
```

Comment:
See my other long comment. The signature result for lists is correct, even though it is intuitive.